src/main/java/org/gaul/s3proxy/junit/rules/S3ProxyRule.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2014-2018 Andrew Gaul <andrew@gaul.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy.junit.rules;
+
+import java.net.URI;
+import java.util.Properties;
+
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.gaul.s3proxy.AuthenticationType;
+import org.gaul.s3proxy.S3Proxy;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Emulates a S3 endpoint.
+ */
+public final class S3ProxyRule extends ExternalResource {
+    private static final Logger logger = LoggerFactory.getLogger(
+        S3ProxyRule.class);
+
+    private static final String ENDPOINT_URL = "http://127.0.0.1";
+
+    private final TemporaryFolder temp = new TemporaryFolder();
+    private final boolean ignoreUnknownHeaders;
+    private final String accessKey;
+    private final String secretKey;
+    private S3Proxy.Builder s3ProxyBuilder;
+    private S3Proxy s3Proxy;
+    private int port;
+
+    public static final class Builder {
```

Comment:
This substantially duplicates `S3Proxy.Builder`.  Can you rework `S3ProxyRule` to take advantage of this?

Reply:
The idea is empower testers to mock the S3 endpoint with little effort and without getting deep into the S3Proxy internals and leveraging on JUnit framework to create and destroy the resources on behalf of the test writers. If we let the rule to take an instance of the `S3Proxy.Builder` instead then the rule lose sense.